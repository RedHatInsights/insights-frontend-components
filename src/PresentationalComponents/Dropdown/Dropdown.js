import React from 'react';
import PropTypes from 'prop-types';
import { generateID } from '../../functions/generateID.js';

import {
    Dropdown as PfDropdown,
    DropdownToggle,
    KebabToggle,
    DropdownDirection,
    DropdownPosition
} from '@patternfly/react-core';

export { DropdownDirection, DropdownPosition } from '@patternfly/react-core';

const Dropdown = ({
    position,
    direction,
    isKebab,
    className,
    title,
    hasArrow,
    isCollapsed,
    onSelect,
    children,
    onToggle,
    widgetId,
    ...props
}) => {
    const Toggle = isKebab ? KebabToggle : DropdownToggle;
    console.warn('Dropdown from FE component shouldn\'t be used anymore. \
Instead use http://patternfly-react.surge.sh/patternfly-4/components/dropdown#Dropdown from PF repository.');
    const extraToggleProps = {
        ...!hasArrow ? { iconComponent: null } : {}
    };
    return (
        <PfDropdown
            widget-type='InsightsDropdown'
            { ...props }
            className={ className }
            widget-id={ widgetId }
            position={ position }
            direction={ direction }
            onSelect={ onSelect }
            isOpen={ !isCollapsed }
            toggle={ <Toggle { ...extraToggleProps } onToggle={ isOpen => onToggle(undefined, !isOpen) }>{ title }</Toggle> }
        >
            { children }
        </PfDropdown>
    );
};

Dropdown.propTypes = {
    position: PropTypes.oneOf(Object.keys(DropdownPosition)),
    direction: PropTypes.oneOf(Object.keys(DropdownDirection)),
    isKebab: PropTypes.bool,
    children: PropTypes.node,
    className: PropTypes.string,
    title: PropTypes.node.isRequired,
    isCollapsed: PropTypes.bool,
    hasArrow: PropTypes.bool,
    onSelect: PropTypes.func,
    onToggle: PropTypes.func,
    widgetId: PropTypes.string
};

Dropdown.defaultProps = {
    widgetId: generateID('Dropdown'),
    onToggle: () => undefined,
    hasArrow: true,
    isCollapsed: true
};

export default Dropdown;
