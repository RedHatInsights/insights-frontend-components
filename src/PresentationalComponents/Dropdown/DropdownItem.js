import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { DropdownItem as PfDropdownItem, DropdownSeparator } from '@patternfly/react-core';

class DropdownItem extends Component {
    render() {
        const {
            className,
            children,
            isDisabled,
            isSeparator,
            ...props
        } = this.props;

        const DropCmp = isSeparator ? DropdownSeparator : PfDropdownItem;

        console.warn('DropdownItem from FE component shouldn\'t be used anymore. \
Instead use http://patternfly-react.surge.sh/patternfly-4/components/dropdown#DropdownItem from PF repository.');
        return (
            <DropCmp
                { ...props }
                isDisabled={ isDisabled }
                className={
                    isSeparator ?
                        className.split(' ').filter(oneClass => oneClass !== 'pf-c-dropdown__menu-item').join(' ') :
                        className
                } >
                { !isSeparator && children }
            </DropCmp>
        );
    }
}

DropdownItem.propTypes = {
    children: PropTypes.node,
    className: PropTypes.string,
    component: PropTypes.node,
    isDisabled: PropTypes.bool
};

DropdownItem.defaultProps = {
    children: null,
    className: '',
    isDisabled: false
};

export default DropdownItem;
