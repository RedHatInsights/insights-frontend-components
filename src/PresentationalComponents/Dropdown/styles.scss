@import '~@patternfly/patternfly-next/components/Dropdown/styles.scss';