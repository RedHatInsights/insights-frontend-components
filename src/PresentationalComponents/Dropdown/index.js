export { default as Dropdown, DropdownPosition, DropdownDirection } from './Dropdown';
export { default as DropdownItem } from './DropdownItem';
import './styles.scss';