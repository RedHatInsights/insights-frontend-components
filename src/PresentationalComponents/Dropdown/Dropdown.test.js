import React from 'react';
import Dropdown, { DropdownPosition, DropdownDirection } from './Dropdown';
import DropdownItem from './DropdownItem';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';

describe('Dropdown component', () => {
    console.warn = jest.fn();
    describe('should render correctly', () => {
        test('no props', () => {
            const wrapper = shallow(<Dropdown widgetId="test" title="Test" />);
            expect(toJson(wrapper)).toMatchSnapshot();
            expect(console.warn.mock.calls.length).toBe(1);
        });

        test('title as dom', () => {
            const wrapper = shallow(<Dropdown widgetId="test" title={<div>something</div>} />);
            expect(toJson(wrapper)).toMatchSnapshot();
        })

        test('no arrow', () => {
            const wrapper = shallow(<Dropdown widgetId="test" hasArrow={false} title="Test" />);
            expect(toJson(wrapper)).toMatchSnapshot();
        });

        Object.values(DropdownPosition).forEach((position) => {
            test(`position - ${position}`, () => {
                const wrapper = shallow(<Dropdown widgetId="test" position={position} title="Test" />);
                expect(toJson(wrapper)).toMatchSnapshot();
            });
        });

        Object.values(DropdownDirection).forEach((direction) => {
            test(`direction - ${direction}`, () => {
                const wrapper = shallow(<Dropdown widgetId="test" direction={direction} title="Test" />);
                expect(toJson(wrapper)).toMatchSnapshot();
            });
        });

        test('is opened', () => {
            const wrapper = shallow(<Dropdown widgetId="test" isCollapsed={false} title="Test" />);
            expect(toJson(wrapper)).toMatchSnapshot();
        });

        test('isKebab', () => {
            const wrapper = shallow(<Dropdown widgetId="test" isKebab title="Test" />);
            expect(toJson(wrapper)).toMatchSnapshot();
        });
    });

    describe('API', () => {
        const onToggle = jest.fn();
        const onSelect = jest.fn();

        it('should call on toggle with false', () => {
            const wrapper = mount(<Dropdown widgetId="test" isKebab title="Test" onToggle={onToggle} />);
            wrapper.find('button.pf-c-dropdown__toggle').first().simulate('click');
            expect(onToggle.mock.calls.length).toBe(1);
            expect(onToggle.mock.calls[0][1]).toBe(false);
        });

        it('should call on toggle with true', () => {
            const wrapper = mount(<Dropdown widgetId="test" isKebab title="Test" onToggle={onToggle} isCollapsed={false} />);
            wrapper.find('button.pf-c-dropdown__toggle').first().simulate('click');
            expect(onToggle.mock.calls.length).toBe(2);
            expect(onToggle.mock.calls[1][1]).toBe(true);
        });

        it('should not call on toggle', () => {
            const wrapper = mount(<Dropdown widgetId="test" isKebab title="Test" />);
            wrapper.find('button.pf-c-dropdown__toggle').first().simulate('click');
            expect(onToggle.mock.calls.length).toBe(2);
        });

        it('should call on select', () => {
            const wrapper = mount(
                <Dropdown widgetId="test" isKebab title="Test" onSelect={onSelect} isCollapsed={false}>
                    <DropdownItem>test</DropdownItem>
                </Dropdown>
            );
            wrapper.find('ul.pf-c-dropdown__menu li').first().simulate('click');
            expect(onSelect.mock.calls.length).toBe(1);
        });

        it('should not call on select', () => {
            const wrapper = mount(
                <Dropdown widgetId="test" isKebab title="Test" isCollapsed={false}>
                    <DropdownItem>test</DropdownItem>
                </Dropdown>
            );
            wrapper.find('ul.pf-c-dropdown__menu li').first().simulate('click');
            expect(onSelect.mock.calls.length).toBe(1);
        });
    });
});
