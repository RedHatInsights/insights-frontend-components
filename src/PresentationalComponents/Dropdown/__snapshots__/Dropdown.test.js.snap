// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown component should render correctly direction - down 1`] = `
<Dropdown
  className=""
  direction="down"
  dropdownItems={Array []}
  isOpen={false}
  isPlain={false}
  onSelect={[Function]}
  position="left"
  toggle={
    <DropdownToggle
      className=""
      iconComponent={[Function]}
      id=""
      isActive={false}
      isFocused={false}
      isHovered={false}
      isOpen={false}
      isPlain={false}
      onToggle={[Function]}
      parentRef={null}
    >
      Test
    </DropdownToggle>
  }
  widget-id="test"
  widget-type="InsightsDropdown"
/>
`;

exports[`Dropdown component should render correctly direction - up 1`] = `
<Dropdown
  className=""
  direction="up"
  dropdownItems={Array []}
  isOpen={false}
  isPlain={false}
  onSelect={[Function]}
  position="left"
  toggle={
    <DropdownToggle
      className=""
      iconComponent={[Function]}
      id=""
      isActive={false}
      isFocused={false}
      isHovered={false}
      isOpen={false}
      isPlain={false}
      onToggle={[Function]}
      parentRef={null}
    >
      Test
    </DropdownToggle>
  }
  widget-id="test"
  widget-type="InsightsDropdown"
/>
`;

exports[`Dropdown component should render correctly is opened 1`] = `
<Dropdown
  className=""
  direction="down"
  dropdownItems={Array []}
  isOpen={true}
  isPlain={false}
  onSelect={[Function]}
  position="left"
  toggle={
    <DropdownToggle
      className=""
      iconComponent={[Function]}
      id=""
      isActive={false}
      isFocused={false}
      isHovered={false}
      isOpen={false}
      isPlain={false}
      onToggle={[Function]}
      parentRef={null}
    >
      Test
    </DropdownToggle>
  }
  widget-id="test"
  widget-type="InsightsDropdown"
/>
`;

exports[`Dropdown component should render correctly isKebab 1`] = `
<Dropdown
  className=""
  direction="down"
  dropdownItems={Array []}
  isOpen={false}
  isPlain={false}
  onSelect={[Function]}
  position="left"
  toggle={
    <Kebab
      aria-label="Actions"
      className=""
      id=""
      isActive={false}
      isDisabled={false}
      isFocused={false}
      isHovered={false}
      isOpen={false}
      isPlain={false}
      onToggle={[Function]}
      parentRef={null}
    >
      Test
    </Kebab>
  }
  widget-id="test"
  widget-type="InsightsDropdown"
/>
`;

exports[`Dropdown component should render correctly no arrow 1`] = `
<Dropdown
  className=""
  direction="down"
  dropdownItems={Array []}
  isOpen={false}
  isPlain={false}
  onSelect={[Function]}
  position="left"
  toggle={
    <DropdownToggle
      className=""
      iconComponent={null}
      id=""
      isActive={false}
      isFocused={false}
      isHovered={false}
      isOpen={false}
      isPlain={false}
      onToggle={[Function]}
      parentRef={null}
    >
      Test
    </DropdownToggle>
  }
  widget-id="test"
  widget-type="InsightsDropdown"
/>
`;

exports[`Dropdown component should render correctly no props 1`] = `
<Dropdown
  className=""
  direction="down"
  dropdownItems={Array []}
  isOpen={false}
  isPlain={false}
  onSelect={[Function]}
  position="left"
  toggle={
    <DropdownToggle
      className=""
      iconComponent={[Function]}
      id=""
      isActive={false}
      isFocused={false}
      isHovered={false}
      isOpen={false}
      isPlain={false}
      onToggle={[Function]}
      parentRef={null}
    >
      Test
    </DropdownToggle>
  }
  widget-id="test"
  widget-type="InsightsDropdown"
/>
`;

exports[`Dropdown component should render correctly position - left 1`] = `
<Dropdown
  className=""
  direction="down"
  dropdownItems={Array []}
  isOpen={false}
  isPlain={false}
  onSelect={[Function]}
  position="left"
  toggle={
    <DropdownToggle
      className=""
      iconComponent={[Function]}
      id=""
      isActive={false}
      isFocused={false}
      isHovered={false}
      isOpen={false}
      isPlain={false}
      onToggle={[Function]}
      parentRef={null}
    >
      Test
    </DropdownToggle>
  }
  widget-id="test"
  widget-type="InsightsDropdown"
/>
`;

exports[`Dropdown component should render correctly position - right 1`] = `
<Dropdown
  className=""
  direction="down"
  dropdownItems={Array []}
  isOpen={false}
  isPlain={false}
  onSelect={[Function]}
  position="right"
  toggle={
    <DropdownToggle
      className=""
      iconComponent={[Function]}
      id=""
      isActive={false}
      isFocused={false}
      isHovered={false}
      isOpen={false}
      isPlain={false}
      onToggle={[Function]}
      parentRef={null}
    >
      Test
    </DropdownToggle>
  }
  widget-id="test"
  widget-type="InsightsDropdown"
/>
`;

exports[`Dropdown component should render correctly title as dom 1`] = `
<Dropdown
  className=""
  direction="down"
  dropdownItems={Array []}
  isOpen={false}
  isPlain={false}
  onSelect={[Function]}
  position="left"
  toggle={
    <DropdownToggle
      className=""
      iconComponent={[Function]}
      id=""
      isActive={false}
      isFocused={false}
      isHovered={false}
      isOpen={false}
      isPlain={false}
      onToggle={[Function]}
      parentRef={null}
    >
      <div>
        something
      </div>
    </DropdownToggle>
  }
  widget-id="test"
  widget-type="InsightsDropdown"
/>
`;
